src/hotspot/share/code/codeBlob.cpp

```diff
@@ -315,6 +316,56 @@ AdapterBlob* AdapterBlob::create(CodeBuffer* cb) {
   return blob;
 }
 
+MethodHandleIntrinsicBlob::MethodHandleIntrinsicBlob(
+  Method* method,
+  int mhi_size,
+  CodeBuffer* code_buffer):
+  RuntimeBlob("MethodHandle intrinsic", code_buffer, sizeof(MethodHandleIntrinsicBlob), mhi_size, CodeOffsets::frame_never_safe, 0, NULL),
+  _method(method) {
+  {
+    debug_only(NoSafepointVerifier nsv;)
+    assert_locked_or_safepoint(CodeCache_lock);
+
+    assert(code_buffer->total_oop_size() == 0, "unexpected oop");
+    assert(code_buffer->total_metadata_size() == 0, "unexpected metadata");
+
+    CodeCache::commit(this);
+  }
+}
+
+void* MethodHandleIntrinsicBlob::operator new(size_t s, int size) throw() {
+  return CodeCache::allocate(size, CodeBlobType::NonNMethod);
+}
+
+MethodHandleIntrinsicBlob* MethodHandleIntrinsicBlob::create(const methodHandle& method,
+  CodeBuffer *code_buffer) {
+  code_buffer->finalize_oop_references(method);
+
+  MethodHandleIntrinsicBlob* mhi = NULL;
+  {
+    MutexLocker mu(CodeCache_lock, Mutex::_no_safepoint_check_flag);
+    int mhi_size = CodeBlob::allocation_size(code_buffer, sizeof(MethodHandleIntrinsicBlob));
```

Comment:
The allocation size could also be computed before taking the code cache lock. BufferBlob also does this for example, but others don't. I think it makes sense to have it outside of the mutex block though, to minimize the time we need to hold the lock. I don't see anything in there that seems to require the lock. (Maybe we should clean up other cases in a followup as well).

Reply:
The issue was [filed](https://bugs.openjdk.org/browse/JDK-8289071) and [fixed](https://github.com/openjdk/jdk/pull/9266).

Reply:
Thanks!